File: scala/src/main/scala/cookbook/RecipeFile.scala
package cookbook

import java.nio.file.{Paths, Path, Files}
import scala.io.Source
import com.typesafe.scalalogging.Logger
import org.slf4j.LoggerFactory

class RecipeFile private extends Recipe {

    val logger = Logger(LoggerFactory getLogger this.getClass)

    def title() = {

        logger debug "# File"

    }

    def cook() = {

        val path = Paths get "src/main/scala/Cooker.scala"

        // 基本情報

        logger debug s"dir=${path.getParent.toAbsolutePath}"
        logger debug s"file=${path.getFileName}"
        logger debug s"size=${Files size path}"

        // 行数カウント

        var src = Source fromFile path.toFile
        try {
            logger debug s"lc=${src.getLines.size}"
        } finally {
            src.close
        }

        src = Source fromFile path.toFile
        try {
            var counter = 0
            for(line <- src.getLines) {
                counter += 1
            }
            logger debug s"lc=${counter}"
        } finally {
            src.close
        }

    }

}

object RecipeFile {

    def apply() = {

        new RecipeFile

    }

    def main(args: Array[String]) {

        val recipe = RecipeFile.apply

        recipe.title
        recipe.cook

    }

}

File: java/src/main/java/cookbook/RecipeFile.java
package cookbook;

import java.nio.file.Paths;
import java.nio.file.Path;
import java.nio.file.Files;
import java.util.function.ToIntFunction;
import java.io.BufferedReader;
import java.io.IOException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RecipeFile implements Recipe {

    private static final Logger LOGGER = LoggerFactory.getLogger(RecipeBase.class);

    public void title() {

        LOGGER.debug("# File");

    }

    public void cook() {

        Path path = Paths.get("src/main/java/Cooker.java");

        // 基本情報

        LOGGER.debug(String.format("dir=%s", path.getParent().toAbsolutePath()));
        LOGGER.debug(String.format("file=%s", path.getFileName()));
        try {
            LOGGER.debug(String.format("size=%d", Files.size(path)));
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage(), ex);
        }

        // 行数カウント

        try {
            LOGGER.debug(String.format("lc=%d", Files.lines(path).count()));
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage(), ex);
        }

        try {
            ToIntFunction<String> one = line -> 1;
            LOGGER.debug(String.format("lc=%d", Files.lines(path).mapToInt(one).sum()));
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage(), ex);
        }

        BufferedReader reader = null;
        try {
            reader = Files.newBufferedReader(path);
            int counter = 0;
            String line;
            while((line = reader.readLine()) != null) {
                counter += 1;
            }
            LOGGER.debug(String.format("lc=%d", counter));
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage(), ex);
        } finally {
            if(reader != null) {
                try {
                    reader.close();
                } catch (IOException ex) {
                    LOGGER.error(ex.getMessage(), ex);
                }
            }
        }

    }

    public static void main(String[] args) throws Exception {

        Recipe recipe = new RecipeBase();

        recipe.title();
        recipe.cook();

    }

}
